### ISSUE TITLE
rustc doesn't warn on irrefutable bindings without initializer values

### ISSUE THREAD

**Issue by jamesmunns** [rust-lang/rust#98515]
Given the following code: [Playground Link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7d288826422903ef34a6f444b4664e01)

```rust
struct Bar {
    some: u32,
    data: [u8; 3]
}

fn takes_bar(bar: &Bar) {
    // this is to throw off the "bar is unused" warning
    if bar.some < 4 {
        return;
    }

    // SHOULD BE:
    // let Bar { some: _some, data } = bar;
    let Bar { some: _some, data };

    data.iter().for_each(|x| println!("{}", x))
}

fn main() {
    let b = Bar { some: 27, data: [1, 2, 3] };
    takes_bar(&b);
}
```

The current output is:

```
   Compiling demo v0.1.0 (/tmp/demo)
error[E0381]: borrow of possibly-uninitialized variable: `data`
  --> src/main.rs:15:5
   |
15 |     data.iter().for_each(|x| println!("{}", x))
   |     ^^^^^^^^^^^ use of possibly-uninitialized `data`

For more information about this error, try `rustc --explain E0381`.
```

Ideally the output should look like:

It might be good to warn on irrefutable bindings without initializers, potentially even suggesting a potential item in scope, if there is one

This reproduces on the current stable (1.61.0) and a relatively recent nightly (rustc 1.63.0-nightly (e71440575 2022-06-02))

**Comment by fmease** [comment]
Today's output:

```
error[E0381]: used binding `data` isn't initialized
  --> src/main.rs:15:5
   |
13 |     let Bar { some: _some, data };
   |                            ---- binding declared here but left uninitialized
14 |
15 |     data.iter().for_each(|x| println!("{}", x))
   |     ^^^^ `data` used here but it isn't initialized
```
